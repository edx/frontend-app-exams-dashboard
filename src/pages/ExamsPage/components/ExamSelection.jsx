import React, { useState } from 'react';
import { useIntl } from '@edx/frontend-platform/i18n';
import {
  MenuItem, SearchField,
  SelectMenu,
} from '@openedx/paragon';
import PropTypes from 'prop-types';

import messages from '../messages';

const ExamSelection = ({ exams, onSelect, isDisabled }) => {
  const { formatMessage } = useIntl();
  const [searchText, setSearchText] = useState('');
  const getMenuItems = () => {
    const menuItems = [
      <MenuItem
        key={0}
        as={SearchField}
        onChange={setSearchText}
        placeholder={formatMessage(messages.examSelectPlaceholder)}
        onSubmit={() => {}}
      />,
    ];
    const examsMatchSearch = exams.filter(exam => (
      exam.name.toLowerCase().includes(searchText.toLowerCase())
    )).map(
      exam => <MenuItem key={exam.id} onClick={() => onSelect(exam.id)}>{exam.name}</MenuItem>,
    );
    const examsNotMatchSearch = exams.filter(exam => !(
      exam.name.toLowerCase().includes(searchText.toLowerCase())
    )).map(
      exam => <MenuItem key={exam.id} disabled>{exam.name}</MenuItem>,
    );
    return menuItems.concat(examsMatchSearch).concat(examsNotMatchSearch);
  };

  return (
    <div data-testid="exam_selection">
      <SelectMenu
        defaultMessage={formatMessage(messages.examSelectDropdownLabel)}
        disabled={isDisabled}
      >
        { getMenuItems() }
      </SelectMenu>
    </div>
  );
};

ExamSelection.propTypes = {
  exams: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
  })).isRequired,
  onSelect: PropTypes.func.isRequired,
  isDisabled: PropTypes.bool,
};

ExamSelection.defaultProps = {
  isDisabled: false,
};

export default ExamSelection;
