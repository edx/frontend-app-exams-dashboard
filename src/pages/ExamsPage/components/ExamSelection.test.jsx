import {
  render, screen,
  waitFor,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import ExamSelection from './ExamSelection';

// nomally mocked for unit tests but required for rendering/snapshots
jest.unmock('react');

const renderWithoutError = (component) => {
  // there is an error thrown inside the SelectMenu component by the
  // popover library. This is a workaround to avoid polluting our output
  // with that error message. Issue: https://github.com/floating-ui/react-popper/issues/350
  jest.spyOn(console, 'error').mockImplementation(() => {});
  const rendered = render(component);
  console.error.mockRestore(); // eslint-disable-line no-console
  return rendered;
};

describe('ExamSelection', () => {
  const defaultExams = [
    { id: 19, name: 'exam1' },
    { id: 27, name: 'exam2' },
    { id: 42, name: 'exam3' },
  ];
  const mockHandleSelectExam = jest.fn();
  test('component matches snapshot', () => {
    expect(renderWithoutError(
      <ExamSelection exams={defaultExams} onSelect={mockHandleSelectExam} />,
    )).toMatchSnapshot();
  });
  it('lists each exam in the dropdown', async () => {
    renderWithoutError(<ExamSelection exams={defaultExams} onSelect={mockHandleSelectExam} />);
    screen.getByText('Select an exam').click();
    defaultExams.forEach((exam) => {
      expect(screen.getByText(exam.name)).toBeInTheDocument();
    });
  });
  it('filters the dropdown by search text', async () => {
    const user = userEvent.setup();
    renderWithoutError(<ExamSelection exams={defaultExams} onSelect={mockHandleSelectExam} />);
    screen.getByText('Select an exam').click();
    const input = screen.getByPlaceholderText('Search for an exam...');
    await user.type(input, 'exam1');
    waitFor(() => {
      expect(screen.queryByRole('link', { name: 'exam1' })).not.toBeDisabled();
      expect(screen.queryByRole('link', { name: 'exam2' })).toBeDisabled();
      expect(screen.queryByRole('link', { name: 'exam3' })).toBeDisabled();
    });
  });
  it('calls onSelect when an exam is selected', () => {
    renderWithoutError(<ExamSelection exams={defaultExams} onSelect={mockHandleSelectExam} />);
    screen.getByText('Select an exam').click();
    screen.getByText('exam2').click();
    expect(mockHandleSelectExam).toHaveBeenCalledTimes(1);
    expect(mockHandleSelectExam).toHaveBeenCalledWith(27);
  });
  it('button disabled when isDisabled is true', () => {
    renderWithoutError(<ExamSelection exams={defaultExams} onSelect={mockHandleSelectExam} isDisabled />);
    expect(screen.getByText('Select an exam')).toBeDisabled();
  });
});
