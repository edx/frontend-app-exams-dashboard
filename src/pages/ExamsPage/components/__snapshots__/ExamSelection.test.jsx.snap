// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExamSelection component matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-testid="exam_selection"
      >
        <div
          class="pgn__menu-select"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="btn btn-outline-primary"
            type="button"
          >
            Select an exam
            <span
              class="pgn__icon btn-icon-after"
            >
              <svg
                aria-hidden="true"
                fill="none"
                focusable="false"
                height="24"
                role="img"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6-1.41-1.41Z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </button>
          <div
            class="pgn__menu-select-popup"
          >
            <div
              style="position: absolute; left: 0px; top: 0px; z-index: 2000;"
            >
              <div
                data-focus-guard="true"
                style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
                tabindex="-1"
              />
              <div
                data-focus-lock-disabled="disabled"
              />
              <div
                data-focus-guard="true"
                style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
                tabindex="-1"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-testid="exam_selection"
    >
      <div
        class="pgn__menu-select"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn btn-outline-primary"
          type="button"
        >
          Select an exam
          <span
            class="pgn__icon btn-icon-after"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="24"
              role="img"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6-1.41-1.41Z"
                fill="currentColor"
              />
            </svg>
          </span>
        </button>
        <div
          class="pgn__menu-select-popup"
        >
          <div
            style="position: absolute; left: 0px; top: 0px; z-index: 2000;"
          >
            <div
              data-focus-guard="true"
              style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
              tabindex="-1"
            />
            <div
              data-focus-lock-disabled="disabled"
            />
            <div
              data-focus-guard="true"
              style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
              tabindex="-1"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
